### PR TITLE
Add EVP_PKEY_todata() and EVP_PKEY_export() functions

### LINKED ISSUE #14657: Should there be an EVP_PKEY_todata()?
We have the API `EVP_PKEY_fromdata()` which creates a new key based on a set of OSSL_PARAMs. It is possible to get param data from an existing key using `EVP_PKEY_get_params`. However this requires you to know up front all the params data that you might want to get and to set up the OSSL_PARAM array accordingly. It would be nice to have an `EVP_PKEY_todata()` function which creates an OSSL_PARAM array for you and populates it with all the relevant parameters required for a subsequent `EVP_PKEY_fromdata()` call to be successful.

### PR DESCRIPTION
**Pull request by slontis**
Fixes #14657

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

This is dependent on #14785

### CODE REVIEW

**Comment by slontis** [comment]
Ignore the first 2 commits which are part of #14785

File: test/evp_extra_test2.c [hunk 1]
```diff
@@ -347,6 +347,287 @@ static int test_d2i_PrivateKey_ex(void) {
...
+                                                      sizeof(dsa_seed));
+    gen_params[1] = OSSL_PARAM_construct_int(OSSL_PKEY_PARAM_FFC_GINDEX,
+                                             &dsa_gindex);
+    gen_params[2] = OSSL_PARAM_construct_int(OSSL_PKEY_PARAM_FFC_PCOUNTER,
+                                             &dsa_pcounter);
+    gen_params[3] = OSSL_PARAM_construct_end();
+
+    if (!TEST_ptr(pkey = d2i_AutoPrivateKey_ex(NULL, &pkeydata, sizeof(dsa_key),
+                                               mainctx, NULL))
+        || !TEST_int_eq(EVP_PKEY_todata(pkey, EVP_PKEY_KEYPAIR, &to_params), 1)
```
Comment (mattcaswell) [inline test/evp_extra_test2.c:586]: Should there be a round trip test here? i.e. confirm that you can recreate an equivalent key by using EVP_PKEY_fromdata() on the returned params?

File: util/libcrypto.num [hunk 2]
```diff
@@ -5347,3 +5346,8 @@ EVP_ASYM_CIPHER_description             ?	3_0_0	EXIST::FUNCTION:
 EVP_KEM_description                     ?	3_0_0	EXIST::FUNCTION:
 EVP_KEYEXCH_description                 ?	3_0_0	EXIST::FUNCTION:
 EVP_KDF_description                     ?	3_0_0	EXIST::FUNCTION:
+OSSL_PARAM_dup                          ?	3_0_0	EXIST::FUNCTION:
+OSSL_PARAM_merge                        ?	3_0_0	EXIST::FUNCTION:
+OSSL_PARAM_free                         ?	3_0_0	EXIST::FUNCTION:
+EVP_PKEY_todata                         ?	3_0_0	EXIST::FUNCTION:
+EVP_PKEY_export                         ?	3_0_0	EXIST::FUNCTION:
```
Comment (mattcaswell) [inline util/libcrypto.num:5353]: There isn't an explicit test of `EVP_PKEY_export`. Do we need one, or is it sufficiently tested via `EVP_PKEY_todata`?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
One thing I see as slightly odd is that with fromdata you have to call fromdata_init first. But not with todata. That's probably correct behaviour but the asymmetry seems strange.

File: test/evp_extra_test2.c (hunk 1 above)
Comment (slontis) [reply test/evp_extra_test2.c:586]: ok..

File: doc/internal/man3/evp_keymgmt_util_export_to_provider.pod [hunk 3]
```diff
@@ -33,7 +33,8 @@ OP_CACHE_ELEM
 =head1 DESCRIPTION
 
 evp_keymgmt_util_export() calls L<evp_keymgmt_export(3)> with the
-I<keymgmt> and I<keydata> from I<pk>.  This is a convenience function.
+I<keymgmt> and I<keydata> from I<pk>.  This is used as a
+helper for L<EVP_PKEY_todata(3)>.
```
Comment (levitte) [inline doc/internal/man3/evp_keymgmt_util_export_to_provider.pod:37]: ... among other things.

File: doc/man3/EVP_PKEY_todata.pod [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+EVP_PKEY_todata() extracts values from a key I<pkey> using the I<selection>.
+I<selection> is described in L<EVP_PKEY_fromdata(3)/Selections>.
+L<OSSL_PARAM_free(3)> should be used to free the returned parameters in
+I<*params>.
+
+EVP_PKEY_export() is similiar to EVP_PKEY_todata() but uses a callback
+I<export_cb> that gets passed the value of I<export_cbarg>.
+See L<openssl-core.h(7)> for more information about the callback. Note that the
+B<OSSL_PARAM> array that is passed to the callback is not persistent after the
+callback returns. Use EVP_PKEY_todata() if persistent data is required.
```
Comment (levitte) [inline doc/man3/EVP_PKEY_todata.pod:30]: s/Use EVP_PKEY_todata() if persistent data is required/You must preserve the items of interest, or use EVP_PKEY_todata() if persistence is required/

File: doc/man3/EVP_PKEY_todata.pod [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+=head1 NOTES
+
+These functions only work with key management methods coming from a provider.
+This is the mirror function to L<EVP_PKEY_fromdata(3)>.
+
+=head1 RETURN VALUES
+
+EVP_PKEY_todata() and EVP_PKEY_export() return 1 for success and 0 or
+a negative value for failure.
```
Comment (levitte) [inline doc/man3/EVP_PKEY_todata.pod:40]: Uhmmm, I don't think that they return any negative value.

File: doc/man7/evp.pod [hunk 6]
```diff
@@ -32,7 +32,8 @@ with a private key of a particular algorithm by using the functions
 described on the L<EVP_PKEY_fromdata(3)> page, or
 new keys can be generated using L<EVP_PKEY_keygen(3)>.
 EVP_PKEYs can be compared using L<EVP_PKEY_cmp(3)>, or printed using
-L<EVP_PKEY_print_private(3)>.
+L<EVP_PKEY_print_private(3)>. L<EVP_PKEY_todata(3)> can be used to convert a
+key back into raw parameter data.
```
Comment (levitte) [inline doc/man7/evp.pod:36]: "raw" is an unfortunate term here...  how about "back into an L<OSSL_PARAM(3)> array"?

**Review by levitte — CHANGES_REQUESTED** [review]

File: doc/man3/EVP_PKEY_todata.pod (hunk 4 above)
Comment (t8m) [reply doc/man3/EVP_PKEY_todata.pod:30]: Maybe copy instead of preserve?

**Comment by levitte** [comment]
> One thing I see as slightly odd is that with fromdata you have to call fromdata_init first. But not with todata. That's probably correct behaviour but the asymmetry seems strange.

Let's not bend over backwards over such things...

File: doc/man3/EVP_PKEY_todata.pod (hunk 4 above)
Comment (levitte) [reply doc/man3/EVP_PKEY_todata.pod:30]: You can't know exactly how they want to do that.  I can imagine have a callback that writes the values to a BIO that's passed as the cb arg.

**Comment by richsalz** [comment]
> Let's not bend over backwards over such things...

:(  Consistency to developers matters a great deal.

File: test/evp_extra_test2.c (hunk 1 above)
Comment (slontis) [reply test/evp_extra_test2.c:586]: Yes - added..

File: util/libcrypto.num (hunk 2 above)
Comment (slontis) [reply util/libcrypto.num:5353]: Added an export test..

File: doc/internal/man3/evp_keymgmt_util_export_to_provider.pod (hunk 3 above)
Comment (slontis) [reply doc/internal/man3/evp_keymgmt_util_export_to_provider.pod:37]: I think this is sufficient unless you want to suggest alternate wording?

**Comment by mattcaswell** [comment]
> :( Consistency to developers matters a great deal.

I think the problem here is "consistency with what"? "fromdata" is essentially very similar to a keygen operation and the "init then generate" pattern is consistent with how normal keygen works. "todata" is not a keygen operation and the "init then generate" pattern doesn't make a lot of sense and there seems no need for "todata" to attempt to be consistent with keygen. This is where I think the inconsistency comes from - in spite of the apparent similarity in names they are different types of function.

**Comment by mattcaswell** [comment]
This looks ok to me, but I guess we need to wait for #14785 to be merged.

**Comment by slontis** [comment]
Rebased now that #14785 is merged..

**Comment by slontis** [comment]
Ready for review now that dependency has been merged.

**Comment by slontis** [comment]
The appveyor issue should be addressed by https://github.com/openssl/openssl/pull/14851.

**Review by mattcaswell — APPROVED** [review]
LGTM

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master.

**Comment by slontis** [comment]
Thanks - I would of got to it today..

**Pull request closed without merging**